### PR TITLE
Add initial gridpoints to proposeGridpoints

### PR DESCRIPTION
**Pull request by jmirabel**
Changes in this PRs:
- Add initial gridpoints to proposeGridpoints
- improve Python bindings of PathParametrizationAlgorithm

Checklists:
- [x] Update HISTORY.md with a single line describing this PR

### CODE REVIEW

**Comment by jmirabel** [comment]
@hungpham2511 gentle reminder.

File: HISTORY.md [hunk 1]
```diff
@@ -4,6 +4,9 @@
 - [python] seidelWrapper solves 1d LP instead of 2d LP when x_min == x_max and solve_1d > 0
 - [python] Fixed planning_utils.py to follow the latest api of toppra.algorithm.ParameterizationAlgorithm
 
+### Changed
+- [cpp] add initial gridpoints to proposeGridpoints method.
```
Comment (hungpham2511) [inline HISTORY.md]:
```suggestion
- [cpp] add initial gridpoints to proposeGridpoints method.
```

File: cpp/src/toppra/geometric_path.cpp [hunk 2]
```diff
@@ -15,8 +15,27 @@ Vectors GeometricPath::eval(const Vector &positions, int order) const {
...
+    for (int i : {0, 1}) {
+      if (std::abs(I[i] - initialGridpoints[i*N]) > TOPPRA_NEARLY_ZERO) {
+        std::ostringstream oss;
+        oss << "initialGridpoints[" << i*N << "] must be " << I[i] << " and not " << initialGridpoints[i*N];
+        throw std::invalid_argument(oss.str());
+      }
+    }
+    for (int i = N; i >= 0; --i)
+      gridpoints.push_front(initialGridpoints[i]);
+  }
```
Comment (hungpham2511) [inline cpp/src/toppra/geometric_path.cpp:40]: `nb_gridpoints` is not necessarily 2 now.

File: cpp/src/toppra/geometric_path.cpp [hunk 3]
```diff
@@ -15,8 +15,27 @@ Vectors GeometricPath::eval(const Vector &positions, int order) const {
...
+    if (initialGridpoints.size() == 1)
+      throw std::invalid_argument("initialGridpoints should be empty or have at least 2 elements");
+    int N = initialGridpoints.size() - 1;
+    for (int i : {0, 1}) {
+      if (std::abs(I[i] - initialGridpoints[i*N]) > TOPPRA_NEARLY_ZERO) {
+        std::ostringstream oss;
+        oss << "initialGridpoints[" << i*N << "] must be " << I[i] << " and not " << initialGridpoints[i*N];
+        throw std::invalid_argument(oss.str());
+      }
+    }
```
Comment (hungpham2511) [inline cpp/src/toppra/geometric_path.cpp:35]: In addition to this check, we should also check that the given elements are monotonically increasing.

**Review by hungpham2511 — COMMENTED** [review]
While here, can you also provide an unit test for the propose gridpoints function?

File: cpp/src/toppra/geometric_path.cpp (hunk 2 above)
Comment (jmirabel) [reply cpp/src/toppra/geometric_path.cpp:40]: ~~I don't understand what you mean.~~

EDIT: Opening the entire file cleared my doubts.

**Comment by jmirabel** [comment]
I updated the code based on the comments. @hungpham2511 could you have another look please ?

**Review by hungpham2511 — APPROVED** [review]
LGTM Thanks!

**Comment by jmirabel** [comment]
I don't understand why CI job `Python package / cpp (3.6, ubuntu-18.04) (pull_request)` gets cancelled...

**Comment by hungpham2511** [comment]
> I don't understand why CI job `Python package / cpp (3.6, ubuntu-18.04) (pull_request)` gets cancelled...

Indeed it's very strange. I am checking with develop branch.

**Comment by hungpham2511** [comment]
CI passes after restarting.

**Comment by jmirabel** [comment]
Thanks.

**Pull request merged**